cf-oscore/src/main/java/org/eclipse/californium/oscore/OscoreOptionDecoder.java

```diff
@@ -0,0 +1,220 @@
+/*******************************************************************************
+ * Copyright (c) 2022 RISE and others.
+ * 
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v2.0
+ * and Eclipse Distribution License v1.0 which accompany this distribution.
+ * 
+ * The Eclipse Public License is available at
+ *    http://www.eclipse.org/legal/epl-v20.html
+ * and the Eclipse Distribution License is available at
+ *    http://www.eclipse.org/org/documents/edl-v10.html.
+ * 
+ * Contributors:
+ *    Rikard Höglund (RISE)
+ *    
+ ******************************************************************************/
+package org.eclipse.californium.oscore;
+
+import java.nio.ByteBuffer;
+import java.util.Arrays;
+
+/**
+ * Class for decoding the bytes of an OSCORE CoAP option.
+ * 
+ * See the structure of the option:
+ * https://datatracker.ietf.org/doc/html/rfc8613#section-6.1
+ * 
+ */
+public class OscoreOptionDecoder {
+
+	private boolean decoded;
+	private byte[] encodedBytes;
+
+	private byte[] idContext;
+	private byte[] partialIV;
+	private byte[] kid;
+
+	private int n;
+	private int k;
+	private int h;
+
+	/**
+	 * Initialize the OSCORE option with a certain array of bytes.
+	 * 
+	 * @param encodedBytes the encoded bytes of the option
+	 */
+	public OscoreOptionDecoder(byte[] encodedBytes) {
+		this.encodedBytes = encodedBytes;
+	}
+
+	/**
+	 * Set the OSCORE option to a certain array of bytes.
+	 * 
+	 * @param encodedBytes the encoded bytes of the option
+	 */
+	public void setBytes(byte[] encodedBytes) {
+		this.encodedBytes = encodedBytes;
+		decoded = false;
+	}
+
+	/**
+	 * Performs the decoding of the option and stores the resulting parameters
+	 * in this object.
+	 */
+	private void decode() {
+		byte[] total = encodedBytes;
+
+		/**
+		 * If the OSCORE option value is a zero length byte array it represents
+		 * a byte array of length 1 with a byte 0x00 See
+		 * https://tools.ietf.org/html/draft-ietf-core-object-security-16#section-2
+		 */
+		if (total.length == 0) {
+			total = new byte[] { 0x00 };
+		}
+
+		byte flagByte = total[0];
+
+		int n = flagByte & 0x07;
+		int k = flagByte & 0x08;
+		int h = flagByte & 0x10;
+
+		byte[] partialIV = null;
+		byte[] kid = null;
+		byte[] kidContext = null;
+		int index = 1;
+
+		// Parsing Partial IV
+		if (n > 0) {
+			partialIV = Arrays.copyOfRange(total, index, index + n);
+			index += n;
+		}
+
+		// Parsing KID Context
+		if (h != 0) {
+			int s = total[index];
+
+			kidContext = Arrays.copyOfRange(total, index + 1, index + 1 + s);
+
+			index += s + 1;
+		}
+
+		// Parsing KID
+		kid = Arrays.copyOfRange(total, index, total.length);
```

Comment:
The encoding depends on the k-flag in the header-byte.
I'm not sure, where "misformed" options are detected, but I would use the flag to perform some validation here.